tidb panic for "IndexMergeProcessWorker: runtime error: invalid memory address or nil pointer dereference"

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

steps please see in :
[utf-go-clustered-index-panic.log](https://github.com/pingcap/tidb/files/12319665/utf-go-clustered-index-panic.log)


### 2. What did you expect to see? (Required)
no panic
### 3. What did you see instead (Required)
```
[2023/08/10 20:54:12.065 +00:00] [ERROR] [index_merge_reader.go:884] ["IndexMergeProcessWorker: runtime error: invalid memory address or nil pointer dereference"] [conn=2636120476]
[2023/08/10 20:54:12.065 +00:00] [ERROR] [misc.go:91] ["panic in the recoverable goroutine"] [r="\"invalid memory address or nil pointer dereference\""] ["stack trace"="[github.com/pingcap/tidb/util.WithRecovery.func1\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/misc.go:93\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:914\nruntime.panicmem\n\t/usr/local/go/src/runtime/panic.go:261\nruntime.sigpanic\n\t/usr/local/go/src/runtime](http://github.com/pingcap/tidb/util.WithRecovery.func1/n/t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/misc.go:93/nruntime.gopanic/n/t/usr/local/go/src/runtime/panic.go:914/nruntime.panicmem/n/t/usr/local/go/src/runtime/panic.go:261/nruntime.sigpanic/n/t/usr/local/go/src/runtime)
```
### 4. What is your TiDB version? (Required)
master
<!-- Paste the output of SELECT tidb_version() -->

``` go
[2023/08/10 20:54:12.065 +00:00] [ERROR] [index_merge_reader.go:884] ["IndexMergeProcessWorker: runtime error: invalid memory address or nil pointer dereference"] [conn=2636120476]
[2023/08/10 20:54:12.065 +00:00] [ERROR] [misc.go:91] ["panic in the recoverable goroutine"] [r="\"invalid memory address or nil pointer dereference\""] ["stack trace"="github.com/pingcap/tidb/util.WithRecovery.func1
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/misc.go:93
runtime.gopanic
	/usr/local/go/src/runtime/panic.go:914
runtime.panicmem
	/usr/local/go/src/runtime/panic.go:261
runtime.sigpanic
	/usr/local/go/src/runtime/signal_unix.go:861
github.com/pingcap/tidb/util/chunk.(*Chunk).GetRow
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/chunk/chunk.go:357
github.com/pingcap/tidb/executor.handleHeap.Less
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/index_merge_reader.go:948
container/heap.up
	/usr/local/go/src/container/heap/heap.go:92
container/heap.Push
	/usr/local/go/src/container/heap/heap.go:53
github.com/pingcap/tidb/executor.(*indexMergeProcessWorker).fetchLoopUnionWithOrderBy
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/index_merge_reader.go:1066
github.com/pingcap/tidb/executor.(*IndexMergeReaderExecutor).startIndexMergeProcessWorker.func1.1
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/index_merge_reader.go:308
github.com/pingcap/tidb/util.WithRecovery
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/misc.go:96
github.com/pingcap/tidb/executor.(*IndexMergeReaderExecutor).startIndexMergeProcessWorker.func1
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/index_merge_reader.go:300"]
[2023/08/10 20:54:12.066 +00:00] [INFO] [conn.go:1082] ["command dispatched failed"] [conn=2636120476] [connInfo="id:2636120476, addr:10.233.96.131:35656 status:10, collation:utf8mb4_general_ci, user:root"] [command=Query] [status="inTxn:0, autocommit:1"] [sql="select /*+  stream_agg() */ avg(   col_208 ) aggCol from (select  /*+ use_index_merge( tbl_34 ) */ * from tbl_34 where not( tbl_34.col_207 in ( '' , '' , '' , '' ) ) or not( tbl_34.col_208 in ( 16095109 , 3208480 , 12732339 ) ) and not( tbl_34.col_207 in ( '' , '' , '' , '' ) ) and not( tbl_34.col_208 between 3590883 and 10132169 ) order by col_207,col_208  ) ordered_tbl  order by aggCol "] [txn_mode=PESSIMISTIC] [timestamp=443469228157173763] [err="IndexMergeProcessWorker: runtime error: invalid memory address or nil pointer dereference
github.com/pingcap/tidb/executor.(*IndexMergeReaderExecutor).startIndexMergeProcessWorker.func1.handleWorkerPanic.func2
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/index_merge_reader.go:883
github.com/pingcap/tidb/util.WithRecovery.func1
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/misc.go:88
runtime.gopanic
	/usr/local/go/src/runtime/panic.go:914
runtime.panicmem
	/usr/local/go/src/runtime/panic.go:261
runtime.sigpanic
	/usr/local/go/src/runtime/signal_unix.go:861
github.com/pingcap/tidb/util/chunk.(*Chunk).GetRow
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/chunk/chunk.go:357
github.com/pingcap/tidb/executor.handleHeap.Less
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/index_merge_reader.go:948
container/heap.up
	/usr/local/go/src/container/heap/heap.go:92
container/heap.Push
	/usr/local/go/src/container/heap/heap.go:53
github.com/pingcap/tidb/executor.(*indexMergeProcessWorker).fetchLoopUnionWithOrderBy
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/index_merge_reader.go:1066
github.com/pingcap/tidb/executor.(*IndexMergeReaderExecutor).startIndexMergeProcessWorker.func1.1
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/index_merge_reader.go:308
github.com/pingcap/tidb/util.WithRecovery
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/misc.go:96
github.com/pingcap/tidb/executor.(*IndexMergeReaderExecutor).startIndexMergeProcessWorker.func1
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/index_merge_reader.go:300
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1650"]
```

--- a/planner/core/find_best_task.go
+++ b/planner/core/find_best_task.go
@@ -1351,6 +1351,13 @@ func (ds *DataSource) convertToPartialIndexScan(prop *property.PhysicalProperty,
 	is := ds.getOriginalPhysicalIndexScan(prop, path, matchProp, false)
 	// TODO: Consider using isIndexCoveringColumns() to avoid another TableRead
 	indexConds := path.IndexFilters
+	if matchProp {
+		if is.Table.GetPartitionInfo() != nil && !is.Index.Global && is.SCtx().GetSessionVars().StmtCtx.UseDynamicPartitionPrune() {
+			is.Columns, is.schema, _ = AddExtraPhysTblIDColumn(is.SCtx(), is.Columns, is.schema)
+		}
+		// Add sort items for index scan for merge-sort operation between partitions.
+		is.ByItems = byItems
+	}
 	if len(indexConds) > 0 {
 		var selectivity float64
 		if path.CountAfterAccess > 0 {
@@ -1365,13 +1372,6 @@ func (ds *DataSource) convertToPartialIndexScan(prop *property.PhysicalProperty,
 		indexPlan := PhysicalSelection{Conditions: indexConds}.Init(is.SCtx(), stats, ds.SelectBlockOffset())
 		indexPlan.SetChildren(is)
 		return indexPlan
-	}
-	if matchProp {
-		if is.Table.GetPartitionInfo() != nil && !is.Index.Global && is.SCtx().GetSessionVars().StmtCtx.UseDynamicPartitionPrune() {
-			is.Columns, is.schema, _ = AddExtraPhysTblIDColumn(is.SCtx(), is.Columns, is.schema)
-		}
-		// Add sort items for index scan for merge-sort operation between partitions.
-		is.ByItems = byItems
 	}
 	indexPlan = is
 	return indexPlan
@@ -1398,6 +1398,12 @@ func (ds *DataSource) convertToPartialTableScan(prop *property.PhysicalProperty,
 		}
 	}
 	ts.filterCondition = newFilterConds
+	if matchProp {
+		if ts.Table.GetPartitionInfo() != nil && ts.SCtx().GetSessionVars().StmtCtx.UseDynamicPartitionPrune() {
+			ts.Columns, ts.schema, _ = AddExtraPhysTblIDColumn(ts.SCtx(), ts.Columns, ts.schema)
+		}
+		ts.ByItems = byItems
+	}
 	if len(ts.filterCondition) > 0 {
 		selectivity, _, err := cardinality.Selectivity(ds.SCtx(), ds.tableStats.HistColl, ts.filterCondition, nil)
 		if err != nil {
@@ -1407,12 +1413,6 @@ func (ds *DataSource) convertToPartialTableScan(prop *property.PhysicalProperty,
 		tablePlan = PhysicalSelection{Conditions: ts.filterCondition}.Init(ts.SCtx(), ts.StatsInfo().ScaleByExpectCnt(selectivity*rowCount), ds.SelectBlockOffset())
 		tablePlan.SetChildren(ts)
 		return tablePlan
-	}
-	if matchProp {
-		if ts.Table.GetPartitionInfo() != nil && ts.SCtx().GetSessionVars().StmtCtx.UseDynamicPartitionPrune() {
-			ts.Columns, ts.schema, _ = AddExtraPhysTblIDColumn(ts.SCtx(), ts.Columns, ts.schema)
-		}
-		ts.ByItems = byItems
 	}
 	tablePlan = ts
 	return tablePlan

--- a/planner/core/issuetest/BUILD.bazel
+++ b/planner/core/issuetest/BUILD.bazel
@@ -10,7 +10,7 @@ go_test(
     data = glob(["testdata/**"]),
     flaky = True,
     race = "on",
-    shard_count = 9,
+    shard_count = 10,
     deps = [
         "//parser",
         "//planner",

--- a/planner/core/issuetest/planner_issue_test.go
+++ b/planner/core/issuetest/planner_issue_test.go
@@ -198,3 +198,17 @@ func TestIssue46083(t *testing.T) {
 	tk.MustExec("CREATE TEMPORARY TABLE v0(v1 int)")
 	tk.MustExec("INSERT INTO v0 WITH ta2 AS (TABLE v0) TABLE ta2 FOR UPDATE OF ta2;")
 }
+
+func TestIssue46005(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("create table tbl_39(col_239 year(4) not null default '2009', primary key(col_239), unique key idx_223(col_239), key idx_224(col_239))")
+	tk.MustExec("insert into tbl_39 values (1994),(1995),(1996),(1997)")
+	tk.HasPlan(
+		"select  /*+ use_index_merge( tbl_39) */ col_239  from tbl_39 where not( tbl_39.col_239 not in ( '1994' ) ) and tbl_39.col_239 not in ( '2004' , '2010' , '2010' ) or not( tbl_39.col_239 <= '1996' ) and not( tbl_39.col_239 between '2026' and '2011' ) order by tbl_39.col_239 limit 382",
+		"IndexMerge",
+	)
+	rs := tk.MustQuery("select  /*+ use_index_merge( tbl_39) */ col_239  from tbl_39 where not( tbl_39.col_239 not in ( '1994' ) ) and tbl_39.col_239 not in ( '2004' , '2010' , '2010' ) or not( tbl_39.col_239 <= '1996' ) and not( tbl_39.col_239 between '2026' and '2011' ) order by tbl_39.col_239 limit 382")
+	rs.Check(testkit.Rows("1994", "1997"))
+}

